docs/proposals/scheduling/20220701-pod-migration-job.md

```diff
@@ -0,0 +1,354 @@
+
+#### Story 3
+
+The Pod migration capability itself can be provided to users as a service. Users can integrate this API in their own systems to achieve safe migration, and are no longer limited to deschedulers.
+
+
+### Basic Migration API
+
+These APIs provide cluster administrators with more fine-grained migration control capabilities, which can better reduce risks.
+
+- Label `scheduling.koordinator.sh/refuse-to-migrate` indicates that it refuses to migrate this Pod. Although K8s assumes that all Pods should be migrated/evicted, there are always some practical exceptions, such as some stateful applications that may have serious problems if migrated, or some applications that are not yet capable of being migrated .
+- `scheduling.koordinator.sh/migration-priority` indicates the order of migration. Although the K8s community has [Pod Deletion Cost #2255](https://github.com/kubernetes/enhancements/issues/2255), it is not a general mechanism. To avoid conflicts with components that use `Pod Deletion Cost`, users can individually mark migration priorities for Pods. The higher the value, the later in the migration order.
+
+
+### Pod Migration Job CRD
+
+In order to support the above user stories, a Custom Resource Definition(CRD) named `PodMigrationJob` is proposed to ensure the migration process safely.  
+
+#### Migration Job Spec
+
+```go
+
+// PodMigrationJob is the Schema for the PodMigrationJob API
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:scope=Cluster
+type PodMigrationJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              PodMigrationJobSpec   `json:"spec,omitempty"`
+	Status            PodMigrationJobStatus `json:"status,omitempty"`
+}
+
+type PodMigrationJobSpec struct {
+	// Paused indicates whether the PodMigrationJob should to work or not.
+	// Default is false
+	// +optional
+	Paused bool `json:"paused,omitempty"`
+
+	// TimeoutInSeconds controls the PodMigrationJob timeout duration.
+	// +optional
+	TimeoutInSeconds *int32 `json:"timeoutInSeconds,omitempty"`
+
+	// Mode represents the operating mode of the Job
+	// Default is PodMigrationJobModeReservationFirst
+	// +optional
+	Mode PodMigrationJobMode `json:"mode,omitempty"`
+
+	// PodRef represents the Pod that be migrated
+	// +required
+	PodRef *v1.ObjectReference `json:"podRef,omitempty"`
+
+	// ReservationOptions defines the Reservation options for migrated Pod
+	// +optional
+	ReservationOptions *PodMigrateReservationOptions `json:"reservationOptions,omitempty"`
+
+	// DeleteOptions defines the deleting options for the migrated Pod and preempted Pods
+	// +optional
+	DeleteOptions *metav1.DeleteOptions `json:"deleteOptions,omitempty"`
+}
+
+type PodMigrationJobMode string
+
+const (
+	PodMigrationJobModeReservationFirst PodMigrationJobMode = "ReservationFirst"
+	PodMigrationJobModeDeleteDirectly   PodMigrationJobMode = "DeleteDirectly"
```

Comment:
eviction ？

Reply:
DeleteDirectly means to evict or delete the Pod directly. The exact delete behavior depends on the configuration. Or should it be described in a more abstract word like `DestroyDirectly`?

Reply:
Should we use to evict for this scene?

Reply:
done